_extras/refactor-2-code-refactoring.md

```diff
@@ -0,0 +1,411 @@
+only needs to understand what parameters to provide, rather
+than anything else that might need to be configured prior to the call. 
+For these reasons, you should try and have as much of the complex, analytical and mathematical 
+code are pure functions.
+
+
+Some parts of a program are inevitably impure.
+Programs need to read input from users, generate a graph, or write results to a file or a database.
+Well designed programs separate complex logic from the necessary impure "glue" code that 
+interacts with users and other systems.
+This way, you have easy-to-read and easy-to-test pure code that contains the complex logic 
+and simplified impure code that reads data from a file or gathers user input. Impure code may 
+be harder to test but, when simplified like this, may only require a handful of tests anyway.
+
+> ## Exercise: Refactoring To Use a Pure Function
+> Refactor the `analyse_data()` function to delegate the data analysis to a new 
+> pure function `compute_standard_deviation_by_day()` and separate it 
+> from the impure code that handles the input and output.
+> The pure function should take in the data, and return the analysis result, as follows:
+> ```python
+> def compute_standard_deviation_by_day(data):
+>   # TODO
+>   return daily_standard_deviation
+> ```
+>> ## Solution
+>> The analysis code will be refactored into a separate function that may look something like:
+>> ```python
+>>def compute_standard_deviation_by_day(data):
+>>    daily_std_list = []
+>>    for dataset in data:
+>>        daily_std = dataset.groupby(dataset.index.date).std()
+>>        daily_std_list.append(daily_std)
+>>
+>>    daily_standard_deviation = pd.concat(daily_std_list)
+>>    return daily_standard_deviation
+>> ```
+>> The `analyse_data()` function now calls the `compute_standard_deviation_by_day()` function, 
+>> while keeping all the logic for reading the data, processing it and showing it in a graph:
+>>```python
+>>def analyse_data(data_dir):
+>>    """Calculate the standard deviation by day between datasets.
+>>
+>>    Gets all the measurement data from the CSV files in the data directory,
+>>    works out the mean for each day, and then graphs the standard deviation
+>>    of these means.
+>>    """
+>>    data_file_paths = glob.glob(os.path.join(data_dir, 'rain_data_2015*.csv'))
+>>    if len(data_file_paths) == 0:
+>>        raise ValueError('No CSV files found in the data directory')
+>>    data = map(models.read_variable_from_csv, data_file_paths)
+>>    daily_standard_deviation = compute_standard_deviation_by_day(data)
+>>
+>>    graph_data = {
+>>        'standard deviation by day': daily_standard_deviation,
+>>    }
+>>    # views.visualize(graph_data)
+>>    return daily_standard_deviation
+>>```
+>> Make sure to re-run the regression test to check this refactoring has not
+>> changed the output of `analyse_data()`.
+> {: .solution}
+{: .challenge}
+
+### MapReduce Data Processing Approach
```

Comment:
Note this talks about "MapReduce", but only has the Mapping exercise and nothing about reducing.

I would add the "reducing" paragraph from 3.4 so that it makes sense. However, I'm not going to talk about reduce in the course because I think it's too much material - will leave that for the students to look into if interested.

Reply:
Honestly, it would be simpler to just replace the MapReduce text with a box containing the following:

### Mapping
`map(f, C)` is a function takes another function `f()` and a collection `C` of data items as inputs.
Calling `map(f, C)` applies the function `f(x)` to every data item `x` in a collection `C`
and returns the resulting values as a new collection of the same size.

This is a simple mapping that takes a list of names and
returns a list of the lengths of those names using the built-in function `len()`:

~~~
name_lengths = map(len, ["Mary", "Isla", "Sam"])
print(list(name_lengths))
~~~
{: .language-python}
~~~
[4, 4, 3]
~~~
{: .output}

Then you can point the reader to the functional programming section to learn more about MapReduce

Reply:
Sure thing - I've adapted the material to replace the mapreduce section with this callout box.